File: backend-services/billing-service/src/main/java/com/gara/billing_service/entity/Invoice.java
package com.gara.billing_service.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "invoices")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Invoice {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true)
    private String invoiceNumber; // Ví dụ: INV-2026-0001

    @Column(nullable = false)
    private String repairOrderNumber; // Liên kết với phiếu sửa chữa bên repair-service

    private Long customerId; // Lấy từ repair order

    private Double totalLaborCost; // Tổng tiền công
    private Double totalPartCost;  // Tổng tiền phụ tùng
    private Double totalAmount;    // Tổng tiền thanh toán (labor + part)

    @Column(nullable = false)
    private String status; // UNPAID, PAID, CANCELLED

    private String paymentMethod; // CASH, TRANSFER, CARD

    private LocalDateTime createdAt;
    private LocalDateTime paidAt;

    @OneToMany(mappedBy = "invoice", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<InvoiceItem> items = new ArrayList<>();

    public void addItem(InvoiceItem item) {
        items.add(item);
        item.setInvoice(this);
    }
}

File: mobile_gara/lib/models/invoice.dart
class InvoiceModel {
  final int id;
  final String invoiceNumber;
  final String repairOrderNumber;
  final int customerId;
  final double totalLaborCost;
  final double totalPartCost;
  final double totalAmount;
  final String status; // UNPAID, PAID, CANCELLED
  final String? paymentMethod;
  final String createdAt;
  final String? paidAt;

  InvoiceModel({
    required this.id,
    required this.invoiceNumber,
    required this.repairOrderNumber,
    required this.customerId,
    required this.totalLaborCost,
    required this.totalPartCost,
    required this.totalAmount,
    required this.status,
    this.paymentMethod,
    required this.createdAt,
    this.paidAt,
  });

  factory InvoiceModel.fromJson(Map<String, dynamic> json) {
    return InvoiceModel(
      id: json['id'] ?? 0,
      invoiceNumber: json['invoiceNumber'] ?? '',
      repairOrderNumber: json['repairOrderNumber'] ?? '',
      customerId: json['customerId'] ?? 0,
      totalLaborCost: (json['totalLaborCost'] ?? 0.0).toDouble(),
      totalPartCost: (json['totalPartCost'] ?? 0.0).toDouble(),
      totalAmount: (json['totalAmount'] ?? 0.0).toDouble(),
      status: json['status'] ?? 'UNPAID',
      paymentMethod: json['paymentMethod'],
      createdAt: json['createdAt'] ?? '',
      paidAt: json['paidAt'],
    );
  }
}
